In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("police.csv")

In [3]:
df

stop_date stop_time  county_name driver_gender  driver_age_raw  \
0      2005-01-02     01:55          NaN             M          1985.0   
1      2005-01-18     08:15          NaN             M          1965.0   
2      2005-01-23     23:15          NaN             M          1972.0   
3      2005-02-20     17:15          NaN             M          1986.0   
4      2005-03-14     10:00          NaN             F          1984.0   
...           ...       ...          ...           ...             ...   
91736  2015-12-31     20:27          NaN             M          1986.0   
91737  2015-12-31     20:35          NaN             F          1982.0   
91738  2015-12-31     20:45          NaN             M          1992.0   
91739  2015-12-31     21:42          NaN             M          1993.0   
91740  2015-12-31     22:46          NaN             M          1959.0   

       driver_age driver_race                   violation_raw  \
0            20.0       White                        Speeding   
1            40.0       White                        Speeding   
2            33.0       White                        Speeding   
3            19.0       White                Call for Service   
4            21.0       White                        Speeding   
...           ...         ...                             ...   
91736        29.0       White                        Speeding   
91737        33.0       White  Equipment/Inspection Violation   
91738        23.0       White         Other Traffic Violation   
91739        22.0       White                        Speeding   
91740        56.0    Hispanic                        Speeding   

              violation  search_conducted search_type   stop_outcome  \
0              Speeding             False         NaN       Citation   
1              Speeding             False         NaN       Citation   
2              Speeding             False         NaN       Citation   
3                 Other             False         NaN  Arrest Driver   
4              Speeding             False         NaN       Citation   
...                 ...               ...         ...            ...   
91736          Speeding             False         NaN        Warning   
91737         Equipment             False         NaN        Warning   
91738  Moving violation             False         NaN        Warning   
91739          Speeding             False         NaN       Citation   
91740          Speeding             False         NaN       Citation   

      is_arrested stop_duration  drugs_related_stop  
0           False      0-15 Min               False  
1           False      0-15 Min               False  
2           False      0-15 Min               False  
3            True     16-30 Min               False  
4           False      0-15 Min               False  
...           ...           ...                 ...  
91736       False      0-15 Min               False  
91737       False      0-15 Min               False  
91738       False      0-15 Min               False  
91739       False      0-15 Min               False  
91740       False      0-15 Min               False  

[91741 rows x 15 columns]

# head()

In [4]:
df.head()  #“head()” function to display the top 5 rows from our data set.

stop_date stop_time  county_name driver_gender  driver_age_raw  \
0  2005-01-02     01:55          NaN             M          1985.0   
1  2005-01-18     08:15          NaN             M          1965.0   
2  2005-01-23     23:15          NaN             M          1972.0   
3  2005-02-20     17:15          NaN             M          1986.0   
4  2005-03-14     10:00          NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

# Shape

In [5]:
df.shape #if we want to see the dimension of our dataset, we can use “shape”

(91741, 15)

# info()

In [6]:
df.info() #if we want to know some more information about our dataset, we can use the “info()”

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           91741 non-null  object 
 1   stop_time           91741 non-null  object 
 2   county_name         0 non-null      float64
 3   driver_gender       86406 non-null  object 
 4   driver_age_raw      86414 non-null  float64
 5   driver_age          86120 non-null  float64
 6   driver_race         86408 non-null  object 
 7   violation_raw       86408 non-null  object 
 8   violation           86408 non-null  object 
 9   search_conducted    91741 non-null  bool   
 10  search_type         3196 non-null   object 
 11  stop_outcome        86408 non-null  object 
 12  is_arrested         86408 non-null  object 
 13  stop_duration       86408 non-null  object 
 14  drugs_related_stop  91741 non-null  bool   
dtypes: bool(2), float64(3), object(10)
memory usage: 9.3+

# to_datetime()

In [7]:
import datetime as dt

In [8]:
df["stop_date"] = pd.to_datetime(df["stop_date"])
df["stop_time"] = pd.to_datetime(df["stop_time"],format="%H:%M").dt.time

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   stop_date           91741 non-null  datetime64[ns]
 1   stop_time           91741 non-null  object        
 2   county_name         0 non-null      float64       
 3   driver_gender       86406 non-null  object        
 4   driver_age_raw      86414 non-null  float64       
 5   driver_age          86120 non-null  float64       
 6   driver_race         86408 non-null  object        
 7   violation_raw       86408 non-null  object        
 8   violation           86408 non-null  object        
 9   search_conducted    91741 non-null  bool          
 10  search_type         3196 non-null   object        
 11  stop_outcome        86408 non-null  object        
 12  is_arrested         86408 non-null  object        
 13  stop_duration       86408 non-null  object    

In [51]:
df

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0     2005-01-02  01:55:00             M          1985.0        20.0   
1     2005-01-18  08:15:00             M          1965.0        40.0   
2     2005-01-23  23:15:00             M          1972.0        33.0   
3     2005-02-20  17:15:00             M          1986.0        19.0   
4     2005-03-14  10:00:00             F          1984.0        21.0   
...          ...       ...           ...             ...         ...   
91736 2015-12-31  20:27:00             M          1986.0        29.0   
91737 2015-12-31  20:35:00             F          1982.0        33.0   
91738 2015-12-31  20:45:00             M          1992.0        23.0   
91739 2015-12-31  21:42:00             M          1993.0        22.0   
91740 2015-12-31  22:46:00             M          1959.0        56.0   

      driver_race                   violation_raw         violation  \
0           White                        Speeding          Speeding   
1           White                        Speeding          Speeding   
2           White                        Speeding          Speeding   
3           White                Call for Service             Other   
4           White                        Speeding          Speeding   
...           ...                             ...               ...   
91736       White                        Speeding          Speeding   
91737       White  Equipment/Inspection Violation         Equipment   
91738       White         Other Traffic Violation  Moving violation   
91739       White                        Speeding          Speeding   
91740    Hispanic                        Speeding          Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
4                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
91736             False         NaN        Warning       False      0-15 Min   
91737             False         NaN        Warning       False      0-15 Min   
91738             False         NaN        Warning       False      0-15 Min   
91739             False         NaN       Citation       False      0-15 Min   
91740             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
91736               False  
91737               False  
91738               False  
91739               False  
91740               False  

[91262 rows x 14 columns]

# isnull()

In [10]:
df.isnull().sum()

stop_date                 0
stop_time                 0
county_name           91741
driver_gender          5335
driver_age_raw         5327
driver_age             5621
driver_race            5333
violation_raw          5333
violation              5333
search_conducted          0
search_type           88545
stop_outcome           5333
is_arrested            5333
stop_duration          5333
drugs_related_stop        0
dtype: int64

# drop()

In [12]:
df.drop(["county_name"],axis=1,inplace=True)

In [13]:
df

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0     2005-01-02  01:55:00             M          1985.0        20.0   
1     2005-01-18  08:15:00             M          1965.0        40.0   
2     2005-01-23  23:15:00             M          1972.0        33.0   
3     2005-02-20  17:15:00             M          1986.0        19.0   
4     2005-03-14  10:00:00             F          1984.0        21.0   
...          ...       ...           ...             ...         ...   
91736 2015-12-31  20:27:00             M          1986.0        29.0   
91737 2015-12-31  20:35:00             F          1982.0        33.0   
91738 2015-12-31  20:45:00             M          1992.0        23.0   
91739 2015-12-31  21:42:00             M          1993.0        22.0   
91740 2015-12-31  22:46:00             M          1959.0        56.0   

      driver_race                   violation_raw         violation  \
0           White                        Speeding          Speeding   
1           White                        Speeding          Speeding   
2           White                        Speeding          Speeding   
3           White                Call for Service             Other   
4           White                        Speeding          Speeding   
...           ...                             ...               ...   
91736       White                        Speeding          Speeding   
91737       White  Equipment/Inspection Violation         Equipment   
91738       White         Other Traffic Violation  Moving violation   
91739       White                        Speeding          Speeding   
91740    Hispanic                        Speeding          Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
4                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
91736             False         NaN        Warning       False      0-15 Min   
91737             False         NaN        Warning       False      0-15 Min   
91738             False         NaN        Warning       False      0-15 Min   
91739             False         NaN       Citation       False      0-15 Min   
91740             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
91736               False  
91737               False  
91738               False  
91739               False  
91740               False  

[91741 rows x 14 columns]

In [52]:
df.shape

(91262, 14)

# describe()

In [14]:
df.describe()

driver_age_raw    driver_age
count    86414.000000  86120.000000
mean      1970.491228     34.011333
std        110.914909     12.738564
min          0.000000     15.000000
25%       1967.000000     23.000000
50%       1980.000000     31.000000
75%       1987.000000     43.000000
max       8801.000000     99.000000

# value_counts()

“value_counts()” function is used to identify the different categories in a feature as well as the count of values per category.

In [15]:
df["search_type"].value_counts()

Incident to Arrest                                          1219
Probable Cause                                               891
Inventory                                                    220
Reasonable Suspicion                                         197
Protective Frisk                                             161
Incident to Arrest,Inventory                                 129
Incident to Arrest,Probable Cause                            106
Probable Cause,Reasonable Suspicion                           75
Incident to Arrest,Inventory,Probable Cause                   34
Incident to Arrest,Protective Frisk                           33
Probable Cause,Protective Frisk                               33
Inventory,Probable Cause                                      22
Incident to Arrest,Reasonable Suspicion                       13
Inventory,Protective Frisk                                    11
Incident to Arrest,Inventory,Protective Frisk                 11
Protective Frisk,Reasonab

# fillna()

Now, as we know that, by using the “isnull()” and “sum()” functions, we can check if our data has any missing values or not. So, now we can see that this “driver_gender” feature has 5335 nan or missing values. So, we will fill up the missing values using the mode(a value that appears most frequently in a data set) of this particular feature using the “fillna()” function.

In [16]:
df["driver_gender"].isnull().sum()

5335

In [17]:
df["driver_gender"].fillna(df["driver_gender"].mode,inplace=True)

In [18]:
df["driver_gender"].isnull().sum()

0

# nunique()

We can use the “nunique()” function to find the no. of unique values in our series or data frame.

In [19]:
df.nunique()

stop_date             3768
stop_time             1436
driver_gender            3
driver_age_raw          97
driver_age              78
driver_race              5
violation_raw           12
violation                6
search_conducted         2
search_type             24
stop_outcome             6
is_arrested              2
stop_duration            5
drugs_related_stop       2
dtype: int64

In [20]:
df["driver_race"].nunique()

5

In [21]:
df.columns

Index(['stop_date', 'stop_time', 'driver_gender', 'driver_age_raw',
       'driver_age', 'driver_race', 'violation_raw', 'violation',
       'search_conducted', 'search_type', 'stop_outcome', 'is_arrested',
       'stop_duration', 'drugs_related_stop'],
      dtype='object')

In [50]:
df["driver_race"].unique()

array(['White', 'Black', 'Asian', nan, 'Hispanic', 'Other'], dtype=object)

# nsmallest() & nlargest()

“nsmallest() & nlargest()” functions are used to obtain “n” no. of rows from our dataset which are lowest or highest respectively.

In [23]:
df.nsmallest(10,"driver_age")

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
10500 2006-09-30  01:30:00             M          1991.0        15.0   
17771 2007-06-11  12:30:00             M          1992.0        15.0   
18357 2007-07-04  00:40:00             F          1992.0        15.0   
25294 2008-04-21  11:19:00             M          1993.0        15.0   
45988 2010-11-10  19:39:00             F          1995.0        15.0   
1516  2005-11-22  20:30:00             F          1989.0        16.0   
1710  2005-11-28  00:30:00             M          1989.0        16.0   
2242  2005-12-17  23:50:00             M          1989.0        16.0   
2243  2005-12-17  23:55:00             M          1989.0        16.0   
2867  2006-01-10  20:20:00             M          1990.0        16.0   

      driver_race            violation_raw         violation  \
10500       Black  Other Traffic Violation  Moving violation   
17771       White                 Speeding          Speeding   
18357       White  Other Traffic Violation  Moving violation   
25294    Hispanic  Other Traffic Violation  Moving violation   
45988       White  Other Traffic Violation  Moving violation   
1516        White                 Speeding          Speeding   
1710        White                 Speeding          Speeding   
2242        White                 Speeding          Speeding   
2243        White                 Speeding          Speeding   
2867        White                 Speeding          Speeding   

       search_conducted         search_type   stop_outcome is_arrested  \
10500             False                 NaN  Arrest Driver        True   
17771             False                 NaN       Citation       False   
18357             False                 NaN  Arrest Driver        True   
25294              True  Incident to Arrest  Arrest Driver        True   
45988             False                 NaN       Citation       False   
1516              False                 NaN       Citation       False   
1710              False                 NaN       Citation       False   
2242              False                 NaN       Citation       False   
2243              False                 NaN       Citation       False   
2867              False                 NaN       Citation       False   

      stop_duration  drugs_related_stop  
10500       30+ Min               False  
17771      0-15 Min               False  
18357      0-15 Min               False  
25294       30+ Min               False  
45988     16-30 Min               False  
1516      16-30 Min               False  
1710       0-15 Min               False  
2242       0-15 Min               False  
2243       0-15 Min               False  
2867       0-15 Min               False

In [24]:
df.nlargest(10,"driver_age")

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
71539 2013-09-08  09:50:00             F          1914.0        99.0   
75037 2014-02-14  10:21:00             M          1920.0        94.0   
76956 2014-04-25  16:40:00             M          1924.0        90.0   
74348 2014-01-18  13:07:00             F          1925.0        89.0   
9829  2006-09-06  12:15:00             M          1918.0        88.0   
50377 2011-06-20  12:05:00             M          1923.0        88.0   
68142 2013-04-06  10:48:00             M          1926.0        87.0   
70699 2013-08-02  14:35:00             M          1926.0        87.0   
74474 2014-01-23  20:30:00             M          1927.0        87.0   
76675 2014-04-15  08:15:00             M          1927.0        87.0   

      driver_race                   violation_raw         violation  \
71539       White         Other Traffic Violation  Moving violation   
75037       White         Other Traffic Violation  Moving violation   
76956       White                        Speeding          Speeding   
74348       White  Equipment/Inspection Violation         Equipment   
9829        White                        Speeding          Speeding   
50377       White                        Speeding          Speeding   
68142       White                        Speeding          Speeding   
70699       White         Other Traffic Violation  Moving violation   
74474       White         Other Traffic Violation  Moving violation   
76675       White                Call for Service             Other   

       search_conducted search_type stop_outcome is_arrested stop_duration  \
71539             False         NaN     Citation       False     16-30 Min   
75037             False         NaN     Citation       False      0-15 Min   
76956             False         NaN      Warning       False      0-15 Min   
74348             False         NaN          N/D       False      0-15 Min   
9829              False         NaN     Citation       False     16-30 Min   
50377             False         NaN      Warning       False      0-15 Min   
68142             False         NaN     Citation       False      0-15 Min   
70699             False         NaN     Citation       False     16-30 Min   
74474             False         NaN     Citation       False      0-15 Min   
76675             False         NaN    No Action       False      0-15 Min   

       drugs_related_stop  
71539               False  
75037               False  
76956               False  
74348               False  
9829                False  
50377               False  
68142               False  
70699               False  
74474               False  
76675               False

# groupby()

In [25]:
import numpy as np

In [55]:
df.groupby("driver_race").agg(np.mean)

driver_age_raw  driver_age  search_conducted  drugs_related_stop
driver_race                                                                  
Asian           1971.259424   33.586928          0.022616            0.004878
Black           1970.837528   33.119240          0.064595            0.016395
Hispanic        1974.867258   31.573561          0.061511            0.013294
Other           1966.941423   34.609244          0.012552            0.004184
White           1969.811161   34.575372          0.028434            0.007681

# get_group()

In [31]:
df.groupby("driver_race").get_group("Asian")

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
10    2005-07-13  16:20:00             M          1979.0        26.0   
20    2005-07-28  12:37:00             F          1979.0        26.0   
37    2005-09-30  14:00:00             M          1962.0        43.0   
67    2005-10-01  23:25:00             M          1974.0        31.0   
96    2005-10-03  02:20:00             F          1950.0        55.0   
...          ...       ...           ...             ...         ...   
91502 2015-12-18  20:18:00             M          1988.0        27.0   
91510 2015-12-18  23:37:00             M          1993.0        22.0   
91539 2015-12-20  09:27:00             M          1962.0        53.0   
91612 2015-12-26  09:07:00             M          1942.0        73.0   
91704 2015-12-30  10:56:00             M          1989.0        26.0   

      driver_race                   violation_raw  violation  \
10          Asian                        Speeding   Speeding   
20          Asian                        Speeding   Speeding   
37          Asian                        Speeding   Speeding   
67          Asian                        Speeding   Speeding   
96          Asian                        Speeding   Speeding   
...           ...                             ...        ...   
91502       Asian                        Speeding   Speeding   
91510       Asian                        Speeding   Speeding   
91539       Asian                        Speeding   Speeding   
91612       Asian                        Speeding   Speeding   
91704       Asian  Equipment/Inspection Violation  Equipment   

       search_conducted search_type stop_outcome is_arrested stop_duration  \
10                False         NaN     Citation       False      0-15 Min   
20                False         NaN     Citation       False      0-15 Min   
37                False         NaN     Citation       False      0-15 Min   
67                False         NaN     Citation       False     16-30 Min   
96                False         NaN     Citation       False      0-15 Min   
...                 ...         ...          ...         ...           ...   
91502             False         NaN     Citation       False      0-15 Min   
91510             False         NaN     Citation       False      0-15 Min   
91539             False         NaN     Citation       False      0-15 Min   
91612             False         NaN     Citation       False      0-15 Min   
91704             False         NaN     Citation       False      0-15 Min   

       drugs_related_stop  
10                  False  
20                  False  
37                  False  
67                  False  
96                  False  
...                   ...  
91502               False  
91510               False  
91539               False  
91612               False  
91704               False  

[2259 rows x 14 columns]

# loc() and iloc()

In [32]:
df.loc[:5,["driver_age","violation"]]

driver_age  violation
0        20.0   Speeding
1        40.0   Speeding
2        33.0   Speeding
3        19.0      Other
4        21.0   Speeding
5        23.0  Equipment

In [33]:
df.loc[(df.driver_age<16)&(df.violation=="Speeding")]

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
17771 2007-06-11  12:30:00             M          1992.0        15.0   

      driver_race violation_raw violation  search_conducted search_type  \
17771       White      Speeding  Speeding             False         NaN   

      stop_outcome is_arrested stop_duration  drugs_related_stop  
17771     Citation       False      0-15 Min               False

In [34]:
df.iloc[:5,:] #first five rows and all the columns

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0 2005-01-02  01:55:00             M          1985.0        20.0       White   
1 2005-01-18  08:15:00             M          1965.0        40.0       White   
2 2005-01-23  23:15:00             M          1972.0        33.0       White   
3 2005-02-20  17:15:00             M          1986.0        19.0       White   
4 2005-03-14  10:00:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted search_type   stop_outcome  \
0          Speeding  Speeding             False         NaN       Citation   
1          Speeding  Speeding             False         NaN       Citation   
2          Speeding  Speeding             False         NaN       Citation   
3  Call for Service     Other             False         NaN  Arrest Driver   
4          Speeding  Speeding             False         NaN       Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False  
2       False      0-15 Min               False  
3        True     16-30 Min               False  
4       False      0-15 Min               False

In [35]:
df.sort_values(by="driver_age")

stop_date stop_time                                      driver_gender  \
18357 2007-07-04  00:40:00                                                  F   
17771 2007-06-11  12:30:00                                                  M   
45988 2010-11-10  19:39:00                                                  F   
10500 2006-09-30  01:30:00                                                  M   
25294 2008-04-21  11:19:00                                                  M   
...          ...       ...                                                ...   
91637 2015-12-27  09:41:00  <bound method Series.mode of 0        M\n1    ...   
91660 2015-12-28  02:28:00  <bound method Series.mode of 0        M\n1    ...   
91674 2015-12-28  12:01:00  <bound method Series.mode of 0        M\n1    ...   
91710 2015-12-30  13:27:00  <bound method Series.mode of 0        M\n1    ...   
91713 2015-12-30  17:07:00  <bound method Series.mode of 0        M\n1    ...   

       driver_age_raw  driver_age driver_race            violation_raw  \
18357          1992.0        15.0       White  Other Traffic Violation   
17771          1992.0        15.0       White                 Speeding   
45988          1995.0        15.0       White  Other Traffic Violation   
10500          1991.0        15.0       Black  Other Traffic Violation   
25294          1993.0        15.0    Hispanic  Other Traffic Violation   
...               ...         ...         ...                      ...   
91637             NaN         NaN         NaN                      NaN   
91660             NaN         NaN         NaN                      NaN   
91674             NaN         NaN         NaN                      NaN   
91710             NaN         NaN         NaN                      NaN   
91713             NaN         NaN         NaN                      NaN   

              violation  search_conducted         search_type   stop_outcome  \
18357  Moving violation             False                 NaN  Arrest Driver   
17771          Speeding             False                 NaN       Citation   
45988  Moving violation             False                 NaN       Citation   
10500  Moving violation             False                 NaN  Arrest Driver   
25294  Moving violation              True  Incident to Arrest  Arrest Driver   
...                 ...               ...                 ...            ...   
91637               NaN             False                 NaN            NaN   
91660               NaN             False                 NaN            NaN   
91674               NaN             False                 NaN            NaN   
91710               NaN             False                 NaN            NaN   
91713               NaN             False                 NaN            NaN   

      is_arrested stop_duration  drugs_related_stop  
18357        True      0-15 Min               False  
17771       False      0-15 Min               False  
45988       False     16-30 Min               False  
10500        True       30+ Min               False  
25294        True       30+ Min               False  
...           ...           ...                 ...  
91637         NaN           NaN               False  
91660         NaN           NaN               False  
91674         NaN           NaN               False  
91710         NaN           NaN               False  
91713         NaN           NaN               False  

[91741 rows x 14 columns]

In [36]:
df[["stop_date","driver_age","driver_race","violation","stop_outcome"]].sort_values(by="driver_age")

stop_date  driver_age driver_race         violation   stop_outcome
18357 2007-07-04        15.0       White  Moving violation  Arrest Driver
17771 2007-06-11        15.0       White          Speeding       Citation
45988 2010-11-10        15.0       White  Moving violation       Citation
10500 2006-09-30        15.0       Black  Moving violation  Arrest Driver
25294 2008-04-21        15.0    Hispanic  Moving violation  Arrest Driver
...          ...         ...         ...               ...            ...
91637 2015-12-27         NaN         NaN               NaN            NaN
91660 2015-12-28         NaN         NaN               NaN            NaN
91674 2015-12-28         NaN         NaN               NaN            NaN
91710 2015-12-30         NaN         NaN               NaN            NaN
91713 2015-12-30         NaN         NaN               NaN            NaN

[91741 rows x 5 columns]

# set_index()

we can use Pandas “set_index()” function to set any of your columns as the index.

In [37]:
df.set_index("stop_date").head()

stop_time driver_gender  driver_age_raw  driver_age driver_race  \
stop_date                                                                    
2005-01-02  01:55:00             M          1985.0        20.0       White   
2005-01-18  08:15:00             M          1965.0        40.0       White   
2005-01-23  23:15:00             M          1972.0        33.0       White   
2005-02-20  17:15:00             M          1986.0        19.0       White   
2005-03-14  10:00:00             F          1984.0        21.0       White   

               violation_raw violation  search_conducted search_type  \
stop_date                                                              
2005-01-02          Speeding  Speeding             False         NaN   
2005-01-18          Speeding  Speeding             False         NaN   
2005-01-23          Speeding  Speeding             False         NaN   
2005-02-20  Call for Service     Other             False         NaN   
2005-03-14          Speeding  Speeding             False         NaN   

             stop_outcome is_arrested stop_duration  drugs_related_stop  
stop_date                                                                
2005-01-02       Citation       False      0-15 Min               False  
2005-01-18       Citation       False      0-15 Min               False  
2005-01-23       Citation       False      0-15 Min               False  
2005-02-20  Arrest Driver        True     16-30 Min               False  
2005-03-14       Citation       False      0-15 Min               False

# duplicated()

We can use the “duplicated()” function to find all the duplicate rows in our dataset. And, then we can remove duplicate values using the drop_duplicates() function

In [38]:
print(df.duplicated().sum())

479


In [39]:
df.drop_duplicates(inplace=True)
print(df.duplicated().sum())

0


# select_dtypes()

We can separate the numerical and categorical features from our data frame and create new ones by using the “select_dtypes()” function and include “np.number” to select numerical columns whereas include “objects” for categorical columns.



In [41]:
numerical_df = df.select_dtypes(include=[np.number])
numerical_df

driver_age_raw  driver_age
0              1985.0        20.0
1              1965.0        40.0
2              1972.0        33.0
3              1986.0        19.0
4              1984.0        21.0
...               ...         ...
91736          1986.0        29.0
91737          1982.0        33.0
91738          1992.0        23.0
91739          1993.0        22.0
91740          1959.0        56.0

[91262 rows x 2 columns]

In [42]:
categorical_df = df.select_dtypes(include="object")
categorical_df

stop_time driver_gender driver_race                   violation_raw  \
0      01:55:00             M       White                        Speeding   
1      08:15:00             M       White                        Speeding   
2      23:15:00             M       White                        Speeding   
3      17:15:00             M       White                Call for Service   
4      10:00:00             F       White                        Speeding   
...         ...           ...         ...                             ...   
91736  20:27:00             M       White                        Speeding   
91737  20:35:00             F       White  Equipment/Inspection Violation   
91738  20:45:00             M       White         Other Traffic Violation   
91739  21:42:00             M       White                        Speeding   
91740  22:46:00             M    Hispanic                        Speeding   

              violation search_type   stop_outcome is_arrested stop_duration  
0              Speeding         NaN       Citation       False      0-15 Min  
1              Speeding         NaN       Citation       False      0-15 Min  
2              Speeding         NaN       Citation       False      0-15 Min  
3                 Other         NaN  Arrest Driver        True     16-30 Min  
4              Speeding         NaN       Citation       False      0-15 Min  
...                 ...         ...            ...         ...           ...  
91736          Speeding         NaN        Warning       False      0-15 Min  
91737         Equipment         NaN        Warning       False      0-15 Min  
91738  Moving violation         NaN        Warning       False      0-15 Min  
91739          Speeding         NaN       Citation       False      0-15 Min  
91740          Speeding         NaN       Citation       False      0-15 Min  

[91262 rows x 9 columns]

# concat()

In [43]:
comb_df = pd.concat([numerical_df,categorical_df],axis=1)

In [44]:
comb_df

driver_age_raw  driver_age stop_time driver_gender driver_race  \
0              1985.0        20.0  01:55:00             M       White   
1              1965.0        40.0  08:15:00             M       White   
2              1972.0        33.0  23:15:00             M       White   
3              1986.0        19.0  17:15:00             M       White   
4              1984.0        21.0  10:00:00             F       White   
...               ...         ...       ...           ...         ...   
91736          1986.0        29.0  20:27:00             M       White   
91737          1982.0        33.0  20:35:00             F       White   
91738          1992.0        23.0  20:45:00             M       White   
91739          1993.0        22.0  21:42:00             M       White   
91740          1959.0        56.0  22:46:00             M    Hispanic   

                        violation_raw         violation search_type  \
0                            Speeding          Speeding         NaN   
1                            Speeding          Speeding         NaN   
2                            Speeding          Speeding         NaN   
3                    Call for Service             Other         NaN   
4                            Speeding          Speeding         NaN   
...                               ...               ...         ...   
91736                        Speeding          Speeding         NaN   
91737  Equipment/Inspection Violation         Equipment         NaN   
91738         Other Traffic Violation  Moving violation         NaN   
91739                        Speeding          Speeding         NaN   
91740                        Speeding          Speeding         NaN   

        stop_outcome is_arrested stop_duration  
0           Citation       False      0-15 Min  
1           Citation       False      0-15 Min  
2           Citation       False      0-15 Min  
3      Arrest Driver        True     16-30 Min  
4           Citation       False      0-15 Min  
...              ...         ...           ...  
91736        Warning       False      0-15 Min  
91737        Warning       False      0-15 Min  
91738        Warning       False      0-15 Min  
91739       Citation       False      0-15 Min  
91740       Citation       False      0-15 Min  

[91262 rows x 11 columns]